packages/runtime-core/src/apiDefineComponent.ts

```diff
@@ -43,41 +44,79 @@ export type DefineComponent<
   PP = PublicProps,
   Props = Readonly<ExtractPropTypes<PropsOrPropOptions>>,
   Defaults = ExtractDefaultPropTypes<PropsOrPropOptions>
-> = ComponentPublicInstanceConstructor<
-  CreateComponentPublicInstance<
-    Props,
-    RawBindings,
-    D,
-    C,
-    M,
-    Mixin,
-    Extends,
-    E,
-    PP & Props,
-    Defaults,
-    true
-  > &
-    Props
-> &
-  ComponentOptionsBase<
-    Props,
-    RawBindings,
-    D,
-    C,
-    M,
-    Mixin,
-    Extends,
-    E,
-    EE,
-    Defaults
-  > &
-  PP
+> =
+  // If props is a class we should ifnore all the process
+  (PropsOrPropOptions extends { prototype: ComponentOptionClass }
```

Comment:
This is the most important part of this PR, this pretty much bypasses our `ComponentPublicInstanceConstructor` to prevent the `Props` type to not be modified